pinot-core/src/main/java/org/apache/pinot/core/common/BlockDocIdSet.java

```diff
@@ -48,12 +49,28 @@ public interface BlockDocIdSet {
    */
   long getNumEntriesScannedInFilter();
 
+  /**
+   * Whether this doc-id set is guaranteed to not match any documents. This can be used for short-circuiting operations.
+   */
+  default boolean isAlwaysFalse() {
+    return false;
+  }
+
+  /**
+   * Whether this doc-id set is guaranteed to match all documents. This can be used for short-circuiting operations.
+   */
+  default boolean isAlwaysTrue() {
+    return false;
+  }
+
   /**
    * For scan-based FilterBlockDocIdSet, pre-scans the documents and returns a non-scan-based FilterBlockDocIdSet.
    */
   default BlockDocIdSet toNonScanDocIdSet() {
     BlockDocIdIterator docIdIterator = iterator();
-
+    if (docIdIterator instanceof EmptyDocIdIterator) {
```

Comment:
Do we need this change? This is prioritizing the always empty filter case, which should be extremely rare